sky/provision/azure/instance.py

```diff
@@ -93,3 +97,155 @@ def cleanup_ports(
+#         client=compute_client.virtual_machines, function_name='list')
+#     for vm in list_virtual_machines(resource_group):
+#         if azure.get_tag(vm.tags,
+#                          TAG_RAY_CLUSTER_NAME) == cluster_name_on_cloud:
+#             if worker_only:
+#                 if azure.get_tag(vm.tags, TAG_RAY_NODE_KIND) != 'worker':
+#                     continue
+#             poller = delete_virtual_machine(resource_group, vm.name)
+#             poller.wait()
+#             if poller.status() != 'Succeeded':
+#                 with ux_utils.print_exception_no_traceback():
+#                     raise ValueError(f'Failed to delete VM {vm.name}: '
+#                                      f'{poller.status()}')
+
+# def _get_vm_ips(network_client, vm, resource_group: str,
+#                 use_internal_ips: bool) -> Tuple[str, str]:
+#     nic_id = vm.network_profile.network_interfaces[0].id
+#     nic_name = nic_id.split("/")[-1]
+#     nic = network_client.network_interfaces.get(
+#         resource_group_name=resource_group,
+#         network_interface_name=nic_name,
+#     )
+#     ip_config = nic.ip_configurations[0]
+
+#     external_ip = None
+#     if not use_internal_ips:
+#         public_ip_id = ip_config.public_ip_address.id
+#         public_ip_name = public_ip_id.split("/")[-1]
+#         public_ip = network_client.public_ip_addresses.get(
+#             resource_group_name=resource_group,
+#             public_ip_address_name=public_ip_name,
+#         )
+#         external_ip = public_ip.ip_address
+
+#     internal_ip = ip_config.private_ip_address
+
+#     return (external_ip, internal_ip)
+
+
+def _get_vm_status(compute_client, vm_name: str, resource_group: str) -> str:
+    instance = compute_client.virtual_machines.instance_view(
+        resource_group_name=resource_group, vm_name=vm_name).as_dict()
+    for status in instance['statuses']:
+        code_state = status['code'].split('/')
+        # It is possible that sometimes the 'code' is empty string, and we
+        # should skip them.
+        if len(code_state) != 2:
+            continue
+        code, state = code_state
+        # skip provisioning status
+        if code == 'PowerState':
+            return state
+    raise ValueError(f'Failed to get status for VM {vm_name}')
+
+
+def _filter_instances(compute_client, filters: Dict[str, str],
+                      resource_group: str) -> List[Any]:
+
+    def match_tags(vm):
+        for k, v in filters.items():
+            if vm.tags.get(k) != v:
+                return False
+        return True
+
+    try:
+        list_virtual_machines = get_azure_sdk_function(
+            client=compute_client.virtual_machines, function_name='list')
+        vms = list_virtual_machines(resource_group_name=resource_group)
+        nodes = list(filter(match_tags, vms))
+    except azure.exceptions().ResourceNotFoundError as e:
+        if 'ResourceGroupNotFound' in str(e):
```

Comment:
Could we have a constant for this? Seems like this is widely used in a lot of PRs including #3032 and #3697.

Reply:
Added. Thanks!